File: app/src/main/java/com/pr0gramm/app/ui/MessageAdapter.java
package com.pr0gramm.app.ui;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.common.collect.Iterables;
import com.pr0gramm.app.Dagger;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.services.UserService;

import java.util.ArrayList;
import java.util.List;


/**
 */
public class MessageAdapter extends RecyclerView.Adapter<MessageAdapter.MessageViewHolder> {
    protected final List<Api.Message> messages;
    private final Context context;
    private final MessageActionListener actionListener;
    private final int itemLayout;

    @Nullable
    private final String userName;
    private MessageView.PointsVisibility pointsVisibility = MessageView.PointsVisibility.CONDITIONAL;

    public MessageAdapter(Context context, List<Api.Message> messages, MessageActionListener actionListener, int layout) {
        this.context = context;
        this.actionListener = actionListener;
        this.messages = new ArrayList<>(messages);
        this.itemLayout = layout;

        UserService userService = Dagger.appComponent(context).userService();
        userName = userService.getName().orNull();

        setHasStableIds(true);
    }

    /**
     * Replace all the messages with the new messages from the given iterable.
     */
    public void setMessages(Iterable<Api.Message> messages) {
        this.messages.clear();
        Iterables.addAll(this.messages, messages);
        notifyDataSetChanged();
    }

    @Override
    public long getItemId(int position) {
        return messages.get(position).id();
    }

    @Override
    public MessageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(itemLayout, parent, false);
        return new MessageViewHolder(view);
    }

    @SuppressWarnings("CodeBlock2Expr")
    @Override
    public void onBindViewHolder(MessageViewHolder holder, int position) {
        Api.Message message = messages.get(position);
        MessageView view = holder.view;
        view.update(message, userName, pointsVisibility);

        if (actionListener != null) {
            view.setOnSenderClickedListener(v -> {
                actionListener.onUserClicked(message.senderId(), message.name());
            });

            boolean isComment = message.itemId() != 0;
            if (isComment) {
                view.setAnswerClickedListener(v -> {
                    actionListener.onAnswerToCommentClicked(message);
                });

                view.setOnClickListener(v -> {
                    actionListener.onCommentClicked(message.itemId(), message.id());
                });
            } else {
                view.setAnswerClickedListener(v -> {
                    actionListener.onAnswerToPrivateMessage(message);
                });
                view.setOnClickListener(null);
            }
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    public void setPointsVisibility(MessageView.PointsVisibility pointsVisibility) {
        this.pointsVisibility = pointsVisibility;
    }

    public static class MessageViewHolder extends RecyclerView.ViewHolder {
        final MessageView view;

        public MessageViewHolder(View itemView) {
            super(itemView);
            view = (MessageView) itemView;
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/MessageAdapter.kt
package com.pr0gramm.app.ui

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.pr0gramm.app.R
import com.pr0gramm.app.api.pr0gramm.Message
import com.pr0gramm.app.api.pr0gramm.MessageType
import com.pr0gramm.app.databinding.ItemDateDividerBinding
import com.pr0gramm.app.ui.fragments.conversation.StringValue
import com.pr0gramm.app.util.inflate
import java.util.Objects

/**
 */
class MessageAdapter(
        private val itemLayout: Int,
        private val actionListener: MessageActionListener,
        private val currentUsername: String?,
        private val paginationController: PaginationController)

    : DelegateAdapter<Any>(DiffCallback()) {

    init {
        delegates += MessageAdapterDelegate()
        delegates += ErrorAdapterDelegate()
        delegates += staticLayoutAdapterDelegate<Loading>(R.layout.feed_hint_loading)
        delegates += staticLayoutAdapterDelegate(R.layout.feed_hint_empty, EmptyValue)

        delegates += run {
            val itemAdapter = Adapters.ForViewBindings(ItemDateDividerBinding::inflate) { (views), item: StringValue ->
                views.text.text = item.text
            }

            Adapters.adapt(itemAdapter) { value -> value as? StringValue }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        paginationController.hit(position, items.size)
        super.onBindViewHolder(holder, position)
    }

    class DiffCallback : DiffUtil.ItemCallback<Any>() {
        override fun areItemsTheSame(oldItem: Any, newItem: Any): Boolean {
            if (oldItem is Message && newItem is Message) {
                return oldItem.id == newItem.id
            }

            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: Any, newItem: Any): Boolean {
            return Objects.equals(oldItem, newItem)
        }
    }

    inner class MessageAdapterDelegate : ListItemTypeAdapterDelegate<Message, Any, MessageViewHolder>(Message::class) {
        override fun onCreateViewHolder(parent: ViewGroup): MessageViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(itemLayout) as MessageView
            return MessageViewHolder(view)
        }

        override fun onBindViewHolder(holder: MessageViewHolder, value: Message) {
            holder.bindTo(value, actionListener, currentUsername)
        }
    }

    object EmptyValue

    open class MessageViewHolder(private val view: MessageView) : RecyclerView.ViewHolder(view) {
        open fun bindTo(message: Message, actionListener: MessageActionListener?, currentUsername: String? = null) {
            view.update(message, currentUsername)

            if (actionListener != null) {
                view.setOnSenderClickedListener {
                    actionListener.onUserClicked(message.senderId, message.name)
                }

                when (message.type) {
                    MessageType.COMMENT -> {
                        view.setAnswerClickedListener(R.string.action_answer) {
                            actionListener.onAnswerToCommentClicked(message)
                        }

                        view.setOnClickListener {
                            actionListener.onCommentClicked(message)
                        }
                    }

                    MessageType.MESSAGE -> {
                        view.setAnswerClickedListener(R.string.action_to_conversation) {
                            actionListener.onAnswerToPrivateMessage(message)
                        }

                        view.setOnClickListener(null)
                    }

                    MessageType.STALK -> {
                        view.clearAnswerClickedListener()

                        view.setOnClickListener {
                            actionListener.onCommentClicked(message)
                        }
                    }

                    else -> {
                        view.clearAnswerClickedListener()
                        view.setOnClickListener(null)
                    }
                }
            }
        }
    }
}

class Loading
